From d981698cd2b4df8bd639724365c4c3cba5e8cff1 Mon Sep 17 00:00:00 2001
From: Tim Carney <tim.carney@cargt.com>
Date: Thu, 28 May 2026 13:31:46 -0500
Subject: [PATCH 27/29] arm64: dts: imx91: Add CARGT 00363 OSM-L SOM and 00365
 carrier board

Add device tree support for the CARGT 00363 OSM-L SOM based on the
MIMX9131CVVXJAB, which is pin-compatible with the MIMX9352CVVXMAB used
on the imx93-cargt-00363 design.

imx91-cargt-00363-osm-som.dts:
  - eMMC (USDHC1, 8-bit HS200)
  - WiFi IW612 (USDHC3 + wake on CCM_CLKO1/GPIO3_IO26)
  - Bluetooth LPUART5 on GPIO_IO00-03 (same balls as i.MX 93)
  - Console LPUART1
  - PMIC PCA9451A on LPI2C1 at 0x25 (interrupt GPIO1_IO10)
  - EEPROM AT24C02 on LPI2C1 at 0x50
  - RTC RX8111 on LPI2C1 at 0x32
  - CMA alloc-ranges lowered to 0x88000000 for 512MB minimum DRAM
  - No media_blk_ctrl (i.MX 91 has no display pipeline)

imx91-cargt-00363-00365.dts:
  - Includes osm-som base DTS
  - Dual RGMII Ethernet: EQOS (ENET1) + FEC (ENET2) via PCA9555 GPIO resets
  - Dual USB OTG with PTN5110 TCPC on LPI2C2
  - CAN1 (PDM_CLK/PDM_BIT_STREAM0) + CAN2 (GPIO_IO25/GPIO_IO27)
  - USDHC2 SD card slot with 3.3V-only, 104MHz limit
  - LPUART2 (Mikrobus) + LPUART6 (GPIO_IO04/05)
  - LPI2C2 (PCA9555 + dual TCPC) + LPI2C3 (GPIO_IO28/29)
  - No display nodes (i.MX 91 is headless)

Signed-off-by: CARGT <engineering@cargt.com>
---
 arch/arm64/boot/dts/freescale/Makefile        |   5 +-
 .../dts/freescale/imx91-cargt-00363-00365.dts | 392 ++++++++++++++++
 .../freescale/imx91-cargt-00363-osm-som.dts   | 442 ++++++++++++++++++
 3 files changed, 837 insertions(+), 2 deletions(-)
 create mode 100644 arch/arm64/boot/dts/freescale/imx91-cargt-00363-00365.dts
 create mode 100644 arch/arm64/boot/dts/freescale/imx91-cargt-00363-osm-som.dts

diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index 4d695bd1ec6a..76ca9f816864 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -452,7 +452,9 @@ dtb-$(CONFIG_ARCH_MXC) += imx93-cargt-00324-00326.dtb \
 				imx93-cargt-00363-00365.dtb \
 				imx93-cargt-00363-00365-glt0557201280is1.dtb \
 				imx93-cargt-00363-00365-glt1011280800is1.dtb \
-				imx93-cargt-00363-00365-glt028240320is1.dtb
+				imx93-cargt-00363-00365-glt028240320is1.dtb \
+				imx91-cargt-00363-osm-som.dtb \
+				imx91-cargt-00363-00365.dtb

 dtb-$(CONFIG_ARCH_MXC) += imx91-9x9-qsb.dtb \
 			  imx91-9x9-qsb-can1.dtb \
diff --git a/arch/arm64/boot/dts/freescale/imx91-cargt-00363-00365.dts b/arch/arm64/boot/dts/freescale/imx91-cargt-00363-00365.dts
new file mode 100644
index 000000000000..5202517c4865
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx91-cargt-00363-00365.dts
@@ -0,0 +1,392 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2022 NXP
+ */
+
+/dts-v1/;
+
+#include <dt-bindings/usb/pd.h>
+#include "imx91-cargt-00363-osm-som.dts"
+
+/ {
+        model = "Cargt i.MX91 00363 OSM-L SOM on 00365 carrier board";
+
+        aliases {
+                rtc0 = &rtc_i2c;
+                rtc1 = &bbnsm_rtc;
+        };
+
+        reg_usdhc2_vmmc: regulator-usdhc2 {
+                pinctrl-names = "default";
+                pinctrl-0 = <&pinctrl_usdhc2_reg>;
+                compatible = "regulator-fixed";
+                regulator-name = "SDIO_A_PWR_EN";
+                regulator-min-microvolt = <3300000>;
+                regulator-max-microvolt = <3300000>;
+                gpio = <&gpio3 7 GPIO_ACTIVE_HIGH>;
+                off-on-delay-us = <12000>;
+                enable-active-high;
+        };
+};
+
+
+&flexcan1 {
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_flexcan1>;
+        status = "okay";
+};
+
+&flexcan2 {
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_flexcan2>;
+        status = "okay";
+};
+
+&lpi2c2 {
+        #address-cells = <1>;
+        #size-cells = <0>;
+        clock-frequency = <400000>;
+        pinctrl-names = "default", "sleep";
+        pinctrl-0 = <&pinctrl_lpi2c2>;
+        pinctrl-1 = <&pinctrl_lpi2c2>;
+        status = "okay";
+
+        pca9555_20: pca9555@20 {
+                compatible = "nxp,pca9555";
+                gpio-controller;
+                #gpio-cells = <2>;
+                reg = <0x20>;
+                gpio-line-names =
+                        "pca9555_20-0", "pca9555_20-1", "pca9555_20-2", "pca9555_20-3",         // 0-3
+                        "pca9555_20-4", "pca9555_20-5", "pca9555_20-6", "pca9555_20-7",         // 4-7
+                        "pca9555_20-8", "pca9555_20-9", "pca9555_20-10","pca9555_20-11",        // 8-11
+                        "pca9555_20-12","pca9555_20-13","pca9555_20-14","pca9555_20-15";        // 12-15
+        };
+
+        ptn5110_51: tcpc@51 {
+                compatible = "nxp,ptn5110";
+                reg = <0x51>;
+                pinctrl-names = "default";
+                pinctrl-0 = <&pinctrl_ptn5110_51>;
+                interrupt-parent = <&gpio2>;
+                interrupts = <9 IRQ_TYPE_LEVEL_LOW>;
+                status = "okay";
+
+                port {
+                        typec1_dr_sw: endpoint {
+                                remote-endpoint = <&usb1_drd_sw>;
+                        };
+                };
+
+                connector {
+                        compatible = "usb-c-connector";
+                        label = "USB-C(A)";
+                        power-role = "dual";
+                        data-role = "dual";
+                        try-power-role = "sink";
+                        source-pdos = <PDO_FIXED(5000, 3000, PDO_FIXED_USB_COMM)>;
+                        sink-pdos = <PDO_FIXED(5000, 3000, PDO_FIXED_USB_COMM)
+                                        PDO_VAR(5000, 20000, 3000)>;
+                        op-sink-microwatt = <15000000>;
+                        self-powered;
+                };
+        };
+
+        ptn5110_52: tcpc@52 {
+                compatible = "nxp,ptn5110";
+                reg = <0x52>;
+                pinctrl-names = "default";
+                pinctrl-0 = <&pinctrl_ptn5110_52>;
+                interrupt-parent = <&gpio2>;
+                interrupts = <10 IRQ_TYPE_LEVEL_LOW>;
+                status = "okay";
+
+                port {
+                        typec2_dr_sw: endpoint {
+                                remote-endpoint = <&usb2_drd_sw>;
+                        };
+                };
+
+                connector {
+                        compatible = "usb-c-connector";
+                        label = "USB-C(C)";
+                        power-role = "dual";
+                        data-role = "dual";
+                        try-power-role = "sink";
+                        source-pdos = <PDO_FIXED(5000, 3000, PDO_FIXED_USB_COMM)>;
+                        sink-pdos = <PDO_FIXED(5000, 3000, PDO_FIXED_USB_COMM)
+                                        PDO_VAR(5000, 20000, 3000)>;
+                        op-sink-microwatt = <15000000>;
+                        self-powered;
+                };
+        };
+};
+
+&lpi2c3 {
+        #address-cells = <1>;
+        #size-cells = <0>;
+        clock-frequency = <400000>;
+        pinctrl-names = "default", "sleep";
+        pinctrl-0 = <&pinctrl_lpi2c3>;
+        pinctrl-1 = <&pinctrl_lpi2c3>;
+        status = "okay";
+ };
+
+&eqos {
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_eqos>;
+        phy-mode = "rgmii-id";
+        phy-handle = <&ethphy1>;
+        snps,reset-gpios = <&pca9555_20 0 GPIO_ACTIVE_LOW>;
+        snps,reset-delays-us = <10 20 200000>;
+        status = "okay";
+
+        mdio {
+                compatible = "snps,dwmac-mdio";
+                #address-cells = <1>;
+                #size-cells = <0>;
+                clock-frequency = <5000000>;
+
+                ethphy1: ethernet-phy@0 {
+                        compatible = "ethernet-phy-ieee802.3-c22";
+                        reg = <0>;
+                        eee-broken-1000t;
+                };
+        };
+};
+
+&fec {
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_fec>;
+        phy-mode = "rgmii-id";
+        phy-handle = <&ethphy2>;
+        fsl,magic-packet;
+        phy-reset-gpios = <&pca9555_20 1 GPIO_ACTIVE_LOW>;
+        phy-reset-duration = <10>;
+        phy-reset-post-delay = <150>;
+        status = "okay";
+
+        mdio {
+                #address-cells = <1>;
+                #size-cells = <0>;
+                clock-frequency = <5000000>;
+
+                ethphy2: ethernet-phy@0 {
+                        compatible = "ethernet-phy-ieee802.3-c22";
+                        reg = <0>;
+                        eee-broken-1000t;
+                };
+        };
+};
+
+&usbotg1 {
+         dr_mode = "otg";
+         hnp-disable;
+         srp-disable;
+         adp-disable;
+         usb-role-switch;
+         disable-over-current;
+         samsung,picophy-pre-emp-curr-control = <3>;
+         samsung,picophy-dc-vol-level-adjust = <7>;
+         status = "okay";
+
+         port {
+                 usb1_drd_sw: endpoint {
+                         remote-endpoint = <&typec1_dr_sw>;
+                 };
+         };
+};
+
+&usbotg2 {
+         dr_mode = "otg";
+         hnp-disable;
+         srp-disable;
+         adp-disable;
+         usb-role-switch;
+         disable-over-current;
+         samsung,picophy-pre-emp-curr-control = <3>;
+         samsung,picophy-dc-vol-level-adjust = <7>;
+         status = "okay";
+
+         port {
+                 usb2_drd_sw: endpoint {
+                         remote-endpoint = <&typec2_dr_sw>;
+                 };
+         };
+};
+
+&usdhc2 {
+         pinctrl-names = "default", "state_100mhz", "state_200mhz";
+         pinctrl-0 = <&pinctrl_usdhc2>, <&pinctrl_usdhc2_gpio>;
+         pinctrl-1 = <&pinctrl_usdhc2_100mhz>, <&pinctrl_usdhc2_gpio>;
+         pinctrl-2 = <&pinctrl_usdhc2_200mhz>, <&pinctrl_usdhc2_gpio>;
+         vmmc-supply = <&reg_usdhc2_vmmc>;
+         cd-gpios = <&gpio3 00 GPIO_ACTIVE_LOW>;
+         bus-width = <4>;
+         status = "okay";
+         no-sdio;
+         no-mmc;
+         max-frequency = <104000000>; /* 104MHz is the max frequency for 00365 */
+         no-1-8-v;
+};
+
+&adc1 {
+        status = "okay";
+};
+
+&lpuart2 { /* Mikrobus UART */
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_uart2>;
+        status = "okay";
+};
+
+&lpuart6 {
+        pinctrl-names = "default";
+        pinctrl-0 = <&pinctrl_uart6>;
+        status = "okay";
+};
+
+&iomuxc {
+
+        pinctrl_eqos: eqosgrp {
+                fsl,pins = <
+                        MX91_PAD_ENET1_MDC__ENET1_MDC                           0x57e
+                        MX91_PAD_ENET1_MDIO__ENET_QOS_MDIO                      0x57e
+                        MX91_PAD_ENET1_RD0__ENET_QOS_RGMII_RD0                  0x57e
+                        MX91_PAD_ENET1_RD1__ENET_QOS_RGMII_RD1                  0x57e
+                        MX91_PAD_ENET1_RD2__ENET_QOS_RGMII_RD2                  0x57e
+                        MX91_PAD_ENET1_RD3__ENET_QOS_RGMII_RD3                  0x57e
+                        MX91_PAD_ENET1_RXC__ENET_QOS_RGMII_RXC                  0x5fe
+                        MX91_PAD_ENET1_RX_CTL__ENET_QOS_RGMII_RX_CTL           0x57e
+                        MX91_PAD_ENET1_TD0__ENET_QOS_RGMII_TD0                  0x57e
+                        MX91_PAD_ENET1_TD1__ENET1_RGMII_TD1                     0x57e
+                        MX91_PAD_ENET1_TD2__ENET_QOS_RGMII_TD2                  0x57e
+                        MX91_PAD_ENET1_TD3__ENET_QOS_RGMII_TD3                  0x57e
+                        MX91_PAD_ENET1_TXC__CCM_ENET_QOS_CLOCK_GENERATE_TX_CLK 0x5fe
+                        MX91_PAD_ENET1_TX_CTL__ENET_QOS_RGMII_TX_CTL           0x57e
+                >;
+        };
+
+        pinctrl_fec: fecgrp {
+                fsl,pins = <
+                        MX91_PAD_ENET2_MDC__ENET2_MDC                           0x57e
+                        MX91_PAD_ENET2_MDIO__ENET2_MDIO                         0x57e
+                        MX91_PAD_ENET2_RD0__ENET2_RGMII_RD0                     0x57e
+                        MX91_PAD_ENET2_RD1__ENET2_RGMII_RD1                     0x57e
+                        MX91_PAD_ENET2_RD2__ENET2_RGMII_RD2                     0x57e
+                        MX91_PAD_ENET2_RD3__ENET2_RGMII_RD3                     0x57e
+                        MX91_PAD_ENET2_RXC__ENET2_RGMII_RXC                     0x5fe
+                        MX91_PAD_ENET2_RX_CTL__ENET2_RGMII_RX_CTL              0x57e
+                        MX91_PAD_ENET2_TD0__ENET2_RGMII_TD0                     0x57e
+                        MX91_PAD_ENET2_TD1__ENET2_RGMII_TD1                     0x57e
+                        MX91_PAD_ENET2_TD2__ENET2_RGMII_TD2                     0x57e
+                        MX91_PAD_ENET2_TD3__ENET2_RGMII_TD3                     0x57e
+                        MX91_PAD_ENET2_TXC__ENET2_RGMII_TXC                     0x5fe
+                        MX91_PAD_ENET2_TX_CTL__ENET2_RGMII_TX_CTL              0x57e
+                >;
+        };
+
+        pinctrl_lpi2c2: lpi2c2grp {
+                fsl,pins = <
+                        MX91_PAD_I2C2_SDA__LPI2C2_SDA                   0x40000b9e
+                        MX91_PAD_I2C2_SCL__LPI2C2_SCL                   0x40000b9e
+                >;
+        };
+
+        pinctrl_lpi2c3: lpi2c3grp {
+                fsl,pins = <
+                        MX91_PAD_GPIO_IO28__LPI2C3_SDA                  0x40000b9e
+                        MX91_PAD_GPIO_IO29__LPI2C3_SCL                  0x40000b9e
+                >;
+        };
+
+        pinctrl_flexcan1: flexcan1grp {
+                fsl,pins = <
+                        MX91_PAD_PDM_CLK__CAN1_TX                               0x139e
+                        MX91_PAD_PDM_BIT_STREAM0__CAN1_RX                       0x139e
+                >;
+        };
+
+        pinctrl_flexcan2: flexcan2grp {
+                fsl,pins = <
+                        MX91_PAD_GPIO_IO25__CAN2_TX                     0x139e
+                        MX91_PAD_GPIO_IO27__CAN2_RX                     0x139e
+                >;
+        };
+
+        pinctrl_usdhc2_reg: usdhc2reggrp {
+                fsl,pins = <
+                        MX91_PAD_SD2_RESET_B__GPIO3_IO7        0x31e /* SDIO_A_PWR_EN */
+                >;
+        };
+
+        pinctrl_usdhc2_gpio: usdhc2gpiogrp {
+                fsl,pins = <
+                        MX91_PAD_SD2_CD_B__GPIO3_IO0           0x31e
+                >;
+        };
+
+        pinctrl_usdhc2: usdhc2grp {
+                fsl,pins = <
+                         MX91_PAD_SD2_CLK__USDHC2_CLK           0x1582
+                         MX91_PAD_SD2_CMD__USDHC2_CMD           0x40001382
+                         MX91_PAD_SD2_DATA0__USDHC2_DATA0       0x40001382
+                         MX91_PAD_SD2_DATA1__USDHC2_DATA1       0x40001382
+                         MX91_PAD_SD2_DATA2__USDHC2_DATA2       0x40001382
+                         MX91_PAD_SD2_DATA3__USDHC2_DATA3       0x40001382
+                         MX91_PAD_SD2_VSELECT__USDHC2_VSELECT   0x51e
+                >;
+         };
+
+        /* need to config the SION for data and cmd pad, refer to ERR052021 */
+        pinctrl_usdhc2_100mhz: usdhc2-100mhzgrp {
+                fsl,pins = <
+                         MX91_PAD_SD2_CLK__USDHC2_CLK           0x158e
+                         MX91_PAD_SD2_CMD__USDHC2_CMD           0x4000138e
+                         MX91_PAD_SD2_DATA0__USDHC2_DATA0       0x4000138e
+                         MX91_PAD_SD2_DATA1__USDHC2_DATA1       0x4000138e
+                         MX91_PAD_SD2_DATA2__USDHC2_DATA2       0x4000138e
+                         MX91_PAD_SD2_DATA3__USDHC2_DATA3       0x4000138e
+                         MX91_PAD_SD2_VSELECT__USDHC2_VSELECT   0x51e
+                >;
+         };
+
+        /* need to config the SION for data and cmd pad, refer to ERR052021 */
+        pinctrl_usdhc2_200mhz: usdhc2-200mhzgrp {
+                fsl,pins = <
+                         MX91_PAD_SD2_CLK__USDHC2_CLK           0x15fe
+                         MX91_PAD_SD2_CMD__USDHC2_CMD           0x400013fe
+                         MX91_PAD_SD2_DATA0__USDHC2_DATA0       0x400013fe
+                         MX91_PAD_SD2_DATA1__USDHC2_DATA1       0x400013fe
+                         MX91_PAD_SD2_DATA2__USDHC2_DATA2       0x400013fe
+                         MX91_PAD_SD2_DATA3__USDHC2_DATA3       0x400013fe
+                         MX91_PAD_SD2_VSELECT__USDHC2_VSELECT   0x51e
+                >;
+         };
+
+        pinctrl_ptn5110_51: ptn5110_51grp {
+                fsl,pins = <
+                        MX91_PAD_GPIO_IO09__GPIO2_IO9                   0x31e
+                >;
+        };
+
+        pinctrl_ptn5110_52: ptn5110_52grp {
+                fsl,pins = <
+                        MX91_PAD_GPIO_IO10__GPIO2_IO10                  0x31e
+                >;
+        };
+
+        pinctrl_uart6: uart6grp {
+                fsl,pins = <
+                        MX91_PAD_GPIO_IO04__LPUART6_TX                  0x31e
+                        MX91_PAD_GPIO_IO05__LPUART6_RX                  0x31e
+                >;
+        };
+
+        pinctrl_uart2: uart2grp {
+                fsl,pins = <
+                        MX91_PAD_UART2_RXD__LPUART2_RX                  0x31e
+                        MX91_PAD_UART2_TXD__LPUART2_TX                  0x31e
+                >;
+        };
+};
diff --git a/arch/arm64/boot/dts/freescale/imx91-cargt-00363-osm-som.dts b/arch/arm64/boot/dts/freescale/imx91-cargt-00363-osm-som.dts
new file mode 100644
index 000000000000..87002be65dbc
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx91-cargt-00363-osm-som.dts
@@ -0,0 +1,442 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2022 NXP
+ */
+
+ /dts-v1/;
+
+ #include <dt-bindings/usb/pd.h>
+ #include "imx91.dtsi"
+
+
+ &ele_fw2 {
+	 memory-region = <&ele_reserved>;
+ };
+
+ / {
+	 model = "Cargt i.MX91 00363 OSM-L SOM";
+	 compatible = "fsl,imx91-11x11-evk", "fsl,imx91";
+
+	 chosen {
+		stdout-path = &lpuart1;
+	};
+
+	reserved-memory {
+		#address-cells = <2>;
+		#size-cells = <2>;
+		ranges;
+
+		linux,cma {
+		   compatible = "shared-dma-pool";
+		   reusable;
+		   alloc-ranges = <0 0x88000000 0 0x18000000>;
+		   size = <0 0xc000000>;
+		   linux,cma-default;
+	   };
+
+		 vdev0vring0: vdev0vring0@a4000000 {
+			reg = <0 0xa4000000 0 0x8000>;
+			no-map;
+		};
+
+		vdev0vring1: vdev0vring1@a4008000 {
+			reg = <0 0xa4008000 0 0x8000>;
+			no-map;
+		};
+
+		vdev1vring0: vdev1vring0@a4000000 {
+			reg = <0 0xa4010000 0 0x8000>;
+			no-map;
+		};
+
+		vdev1vring1: vdev1vring1@a4018000 {
+			reg = <0 0xa4018000 0 0x8000>;
+			no-map;
+		};
+
+	   rsc_table: rsc-table@2021e000 {
+			reg = <0 0x2021e000 0 0x1000>;
+			no-map;
+		};
+
+		vdevbuffer: vdevbuffer@a4020000 {
+			compatible = "shared-dma-pool";
+			reg = <0 0xa4020000 0 0x100000>;
+			no-map;
+		};
+
+		ele_reserved: ele-reserved@a4120000 {
+			compatible = "shared-dma-pool";
+			reg = <0 0xa4120000 0 0x100000>;
+			no-map;
+		};
+	};
+
+	reg_usdhc3_vmmc: regulator-usdhc3 {
+		pinctrl-names = "default";
+		pinctrl-0 = <&pinctrl_wifi_reg>;
+		compatible = "regulator-fixed";
+		regulator-name = "WL_REG_ON";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		gpio = <&gpio2 6 GPIO_ACTIVE_HIGH>; /* WL_REG_ON connects to PDn signal on IW612 */
+		/*
+		 * IW612 wifi chip needs more delay than other wifi chips to complete
+		 * the host interface initialization after power up, otherwise the
+		 * internal state of IW612 may be unstable, resulting in the failure of
+		 * the SDIO3.0 switch voltage.
+		 */
+		startup-delay-us = <20000>;
+		enable-active-high;
+	};
+
+	reg_vref_1v8: regulator-adc-vref {
+		compatible = "regulator-fixed";
+		regulator-name = "vref_1v8";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+	};
+ };
+
+ &gpio1{
+	gpio-line-names =
+	"gpio1-0","gpio1-1","gpio1-2","gpio1-3", 		// 0-3
+	"gpio1-4","gpio1-5","gpio1-6","gpio1-7", 		// 4-7
+	"gpio1-8","gpio1-9","gpio1-10","gpio1-11",	 	// 8-11
+	"gpio1-12","gpio1-13","gpio1-14","gpio1-15", 	// 12-15
+	"gpio1-16","gpio1-17","gpio1-18","gpio1-19", 	// 16-19
+	"gpio1-20","gpio1-21","gpio1-22","gpio1-23", 	// 20-23
+	"gpio1-24","gpio1-25","gpio1-26","gpio1-27", 	// 24-27
+	"gpio1-28","gpio1-29","gpio1-30","gpio1-31"; 	// 28-31
+};
+
+&gpio2{
+	gpio-line-names =
+	"gpio2-0","gpio2-1","gpio2-2","gpio2-3", 		// 0-3
+	"gpio2-4","gpio2-5","gpio2-6","gpio2-7", 		// 4-7
+	"gpio2-8","gpio2-9","gpio2-10","gpio2-11",	 	// 8-11
+	"gpio2-12","gpio2-13","gpio2-14","gpio2-15", 	// 12-15
+	"gpio2-16","gpio2-17","gpio2-18","gpio2-19", 	// 16-19
+	"gpio2-20","gpio2-21","gpio2-22","gpio2-23", 	// 20-23
+	"gpio2-24","gpio2-25","gpio2-26","gpio2-27", 	// 24-27
+	"gpio2-28","gpio2-29","gpio2-30","gpio2-31"; 	// 28-31
+};
+
+&gpio3{
+	gpio-line-names =
+	"gpio3-0","gpio3-1","gpio3-2","gpio3-3", 		// 0-3
+	"gpio3-4","gpio3-5","gpio3-6","gpio3-7", 		// 4-7
+	"gpio3-8","gpio3-9","gpio3-10","gpio3-11",	 	// 8-11
+	"gpio3-12","gpio3-13","gpio3-14","gpio3-15", 	// 12-15
+	"gpio3-16","gpio3-17","gpio3-18","gpio3-19", 	// 16-19
+	"gpio3-20","gpio3-21","gpio3-22","gpio3-23", 	// 20-23
+	"gpio3-24","gpio3-25","gpio3-26","gpio3-27", 	// 24-27
+	"gpio3-28","gpio3-29","gpio3-30","gpio3-31"; 	// 28-31
+};
+
+&gpio4{
+	gpio-line-names =
+	"gpio4-0","gpio4-1","gpio4-2","gpio4-3", 		// 0-3
+	"gpio4-4","gpio4-5","gpio4-6","gpio4-7", 		// 4-7
+	"gpio4-8","gpio4-9","gpio4-10","gpio4-11",	 	// 8-11
+	"gpio4-12","gpio4-13","gpio4-14","gpio4-15", 	// 12-15
+	"gpio4-16","gpio4-17","gpio4-18","gpio4-19", 	// 16-19
+	"gpio4-20","gpio4-21","gpio4-22","gpio4-23", 	// 20-23
+	"gpio4-24","gpio4-25","gpio4-26","gpio4-27", 	// 24-27
+	"gpio4-28","gpio4-29","gpio4-30","gpio4-31"; 	// 28-31
+};
+
+&lpm {
+	 soc-supply = <&buck1>;
+	 status = "okay";
+ };
+
+
+ &lpi2c1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	clock-frequency = <400000>;
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&pinctrl_lpi2c1>;
+	pinctrl-1 = <&pinctrl_lpi2c1>;
+	status = "okay";
+
+	/* eeprom on som */
+	eeprom@50 {
+		compatible = "atmel,24c02";
+		reg = <0x50>;
+		pagesize = <16>;
+	};
+
+	rtc_i2c: rtc@32 {
+		compatible ="epson,rx8111";
+		reg = <0x32>;
+	};
+
+	pmic@25 {
+		compatible = "nxp,pca9451a";
+		reg = <0x25>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&pinctrl_pmic>;
+		interrupt-parent = <&gpio1>;
+		interrupts = <10 IRQ_TYPE_EDGE_FALLING>;
+
+		regulators {
+			buck1: BUCK1 {
+				regulator-name = "BUCK1";
+				regulator-min-microvolt = <650000>;
+				regulator-max-microvolt = <2237500>;
+				regulator-boot-on;
+				regulator-always-on;
+				regulator-ramp-delay = <3125>;
+			};
+
+			buck2: BUCK2 {
+				regulator-name = "BUCK2";
+				regulator-min-microvolt = <600000>;
+				regulator-max-microvolt = <2187500>;
+				regulator-boot-on;
+				regulator-always-on;
+				regulator-ramp-delay = <3125>;
+			};
+
+			buck4: BUCK4{
+				regulator-name = "BUCK4";
+				regulator-min-microvolt = <600000>;
+				regulator-max-microvolt = <3400000>;
+				regulator-boot-on;
+				regulator-always-on;
+			};
+
+			buck5: BUCK5{
+				regulator-name = "BUCK5";
+				regulator-min-microvolt = <600000>;
+				regulator-max-microvolt = <3400000>;
+				regulator-boot-on;
+				regulator-always-on;
+			};
+
+			buck6: BUCK6 {
+				regulator-name = "BUCK6";
+				regulator-min-microvolt = <600000>;
+				regulator-max-microvolt = <3400000>;
+				regulator-boot-on;
+				regulator-always-on;
+			};
+
+			ldo1: LDO1 {
+				regulator-name = "LDO1";
+				regulator-min-microvolt = <1600000>;
+				regulator-max-microvolt = <3300000>;
+				regulator-boot-on;
+				regulator-always-on;
+			};
+
+			ldo4: LDO4 {
+				regulator-name = "LDO4";
+				regulator-min-microvolt = <800000>;
+				regulator-max-microvolt = <3300000>;
+				regulator-boot-on;
+				regulator-always-on;
+			};
+
+			ldo5: LDO5 {
+				regulator-name = "LDO5";
+				regulator-min-microvolt = <1800000>;
+				regulator-max-microvolt = <3300000>;
+				regulator-boot-on;
+				regulator-always-on;
+			};
+		};
+	};
+ };
+
+
+ &lpuart1 { /* console */
+	 pinctrl-names = "default";
+	 pinctrl-0 = <&pinctrl_uart1>;
+	 status = "okay";
+ };
+
+&lpuart5 { /* bluetooth */
+	pinctrl-names = "default";
+	uart-has-rtscts;
+	pinctrl-0 = <&pinctrl_uart5>;
+	status = "okay";
+
+	bluetooth {
+		compatible = "nxp,88w8997-bt";
+		fw-init-baudrate = <3000000>;
+		max-speed = <3000000>;
+	};
+ };
+
+ &mu1 {
+	 status = "okay";
+ };
+
+ &mu2 {
+	 status = "okay";
+ };
+
+ &usdhc1 {
+	 pinctrl-names = "default", "state_100mhz", "state_200mhz";
+	 pinctrl-0 = <&pinctrl_usdhc1>;
+	 pinctrl-1 = <&pinctrl_usdhc1>;
+	 pinctrl-2 = <&pinctrl_usdhc1>;
+	 bus-width = <8>;
+	 non-removable;
+	 status = "okay";
+ };
+
+&usdhc3 {
+	pinctrl-names = "default", "state_100mhz", "state_200mhz", "sleep";
+	pinctrl-0 = <&pinctrl_usdhc3>, <&pinctrl_usdhc3_wlan>;
+	pinctrl-1 = <&pinctrl_usdhc3_100mhz>, <&pinctrl_usdhc3_wlan>;
+	pinctrl-2 = <&pinctrl_usdhc3_200mhz>, <&pinctrl_usdhc3_wlan>;
+	pinctrl-3 = <&pinctrl_usdhc3_sleep>, <&pinctrl_usdhc3_wlan>;
+	vmmc-supply = <&reg_usdhc3_vmmc>;
+	bus-width = <4>;
+	keep-power-in-suspend;
+	non-removable;
+	wakeup-source;
+	status = "okay";
+
+	wifi_wake_host {
+		compatible = "nxp,wifi-wake-host";
+		interrupt-parent = <&gpio3>;
+		interrupts = <26 IRQ_TYPE_LEVEL_LOW>;
+		interrupt-names = "host-wake";
+	};
+};
+
+&usbotg1 {
+	dr_mode = "otg";
+	disable-over-current;
+	samsung,picophy-pre-emp-curr-control = <3>;
+	samsung,picophy-dc-vol-level-adjust = <7>;
+	status = "okay";
+};
+
+&adc1 {
+	vref-supply = <&reg_vref_1v8>;
+	status = "disabled";
+};
+
+&wdog3 {
+	status = "okay";
+};
+
+ &iomuxc {
+
+	 pinctrl_lpi2c1: lpi2c1grp {
+		fsl,pins = <
+			MX91_PAD_I2C1_SCL__LPI2C1_SCL			0x40000b9e
+			MX91_PAD_I2C1_SDA__LPI2C1_SDA			0x40000b9e
+		>;
+	 };
+
+	 pinctrl_uart1: uart1grp {
+		fsl,pins = <
+			MX91_PAD_UART1_RXD__LPUART1_RX			0x31e
+			MX91_PAD_UART1_TXD__LPUART1_TX			0x31e
+		>;
+	};
+
+	pinctrl_uart5: uart5grp {
+		fsl,pins = <
+			MX91_PAD_GPIO_IO00__LPUART5_TX			0x31e
+			MX91_PAD_GPIO_IO01__LPUART5_RX			0x31e
+			MX91_PAD_GPIO_IO02__LPUART5_CTS_B		0x31e
+			MX91_PAD_GPIO_IO03__LPUART5_RTS_B		0x31e
+		>;
+	};
+
+	pinctrl_usdhc1: usdhc1grp {
+		 fsl,pins = <
+			 MX91_PAD_SD1_CLK__USDHC1_CLK			0x159e
+			 MX91_PAD_SD1_CMD__USDHC1_CMD			0x139e
+			 MX91_PAD_SD1_DATA0__USDHC1_DATA0		0x139e
+			 MX91_PAD_SD1_DATA1__USDHC1_DATA1		0x139e
+			 MX91_PAD_SD1_DATA2__USDHC1_DATA2		0x139e
+			 MX91_PAD_SD1_DATA3__USDHC1_DATA3		0x139e
+			 MX91_PAD_SD1_DATA4__USDHC1_DATA4		0x139e
+			 MX91_PAD_SD1_DATA5__USDHC1_DATA5		0x139e
+			 MX91_PAD_SD1_DATA6__USDHC1_DATA6		0x139e
+			 MX91_PAD_SD1_DATA7__USDHC1_DATA7		0x139e
+			 MX91_PAD_SD1_STROBE__USDHC1_STROBE		0x159e
+		 >;
+	 };
+
+	/* need to config the SION for data and cmd pad, refer to ERR052021 */
+	pinctrl_usdhc3: usdhc3grp {
+		fsl,pins = <
+			MX91_PAD_SD3_CLK__USDHC3_CLK			0x1582
+			MX91_PAD_SD3_CMD__USDHC3_CMD			0x40001382
+			MX91_PAD_SD3_DATA0__USDHC3_DATA0		0x40001382
+			MX91_PAD_SD3_DATA1__USDHC3_DATA1		0x40001382
+			MX91_PAD_SD3_DATA2__USDHC3_DATA2		0x40001382
+			MX91_PAD_SD3_DATA3__USDHC3_DATA3		0x40001382
+		>;
+	};
+
+	/* need to config the SION for data and cmd pad, refer to ERR052021 */
+	pinctrl_usdhc3_100mhz: usdhc3-100mhzgrp {
+		fsl,pins = <
+			MX91_PAD_SD3_CLK__USDHC3_CLK			0x158e
+			MX91_PAD_SD3_CMD__USDHC3_CMD			0x4000138e
+			MX91_PAD_SD3_DATA0__USDHC3_DATA0		0x4000138e
+			MX91_PAD_SD3_DATA1__USDHC3_DATA1		0x4000138e
+			MX91_PAD_SD3_DATA2__USDHC3_DATA2		0x4000138e
+			MX91_PAD_SD3_DATA3__USDHC3_DATA3		0x4000138e
+		>;
+	};
+
+	/* need to config the SION for data and cmd pad, refer to ERR052021 */
+	pinctrl_usdhc3_200mhz: usdhc3-200mhzgrp {
+		fsl,pins = <
+			MX91_PAD_SD3_CLK__USDHC3_CLK			0x15be
+			MX91_PAD_SD3_CMD__USDHC3_CMD			0x400013be
+			MX91_PAD_SD3_DATA0__USDHC3_DATA0		0x400013be
+			MX91_PAD_SD3_DATA1__USDHC3_DATA1		0x400013be
+			MX91_PAD_SD3_DATA2__USDHC3_DATA2		0x400013be
+			MX91_PAD_SD3_DATA3__USDHC3_DATA3		0x400013be
+		>;
+	};
+
+	pinctrl_usdhc3_sleep: usdhc3grpsleep {
+		fsl,pins = <
+			MX91_PAD_SD3_CLK__GPIO3_IO20			0x31e
+			MX91_PAD_SD3_CMD__GPIO3_IO21			0x31e
+			MX91_PAD_SD3_DATA0__GPIO3_IO22			0x31e
+			MX91_PAD_SD3_DATA1__GPIO3_IO23			0x31e
+			MX91_PAD_SD3_DATA2__GPIO3_IO24			0x31e
+			MX91_PAD_SD3_DATA3__GPIO3_IO25			0x31e
+		>;
+	};
+
+	pinctrl_usdhc3_wlan: usdhc3wlangrp {
+		fsl,pins = <
+			MX91_PAD_CCM_CLKO1__GPIO3_IO26			0x31e
+		>;
+	};
+
+	pinctrl_rtc: rtcgrp {
+		fsl,pins = <
+			MX91_PAD_GPIO_IO07__GPIO2_IO7			0x31e /* RTC_nINT */
+		>;
+	 };
+
+	pinctrl_wifi_reg: wifireggrp {
+		fsl,pins = <
+			MX91_PAD_GPIO_IO06__GPIO2_IO6			0x31e /* WL_REG_ON */
+		>;
+	};
+
+	pinctrl_pmic: pmicgrp {
+		fsl,pins = <
+			MX91_PAD_PDM_BIT_STREAM1__GPIO1_IO10	0x31e /* PMIC_IRQ_B */
+		>;
+	};
+
+};
-- 
2.34.1

